src/main/java/org/esbtools/auth/util/EnvironmentUtils.java

```diff
@@ -0,0 +1,99 @@
+package org.esbtools.auth.util;
+
+import java.util.Arrays;
+import java.util.List;
+
+import javax.naming.NamingException;
+import javax.naming.directory.NoSuchAttributeException;
+import javax.naming.ldap.LdapName;
+import javax.naming.ldap.Rdn;
+
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EnvironmentUtils {
+
+    private final Logger LOGGER = LoggerFactory.getLogger(EnvironmentUtils.class);
+
+    public static final String ENVIRONMENT_SEPARATOR = ",";
+    public static final String LOCATION = "l";
+    public static final String OU = "ou";
+
+    private final String environment;
+    private final String allAccessOu;
+
+    public String getEnvironment() {
+        return environment;
+    }
+
+    public String getAllAccessOu() {
+        return allAccessOu;
+    }
+
+    public EnvironmentUtils(String environment) {
+        this(environment, null);
+    }
+
+    public EnvironmentUtils(String environment, String allAccessOu) {
+        if (environment == null) {
+            throw new NullPointerException("environment cannot be null");
+        }
+
+        this.environment = environment;
+        this.allAccessOu = allAccessOu;
+    }
+
+    public void validateEnvironment(String certificatePrincipal) throws NamingException {
+
+        String ou = getLDAPAttribute(certificatePrincipal, OU);
+        LOGGER.debug("OU from certificate: ", ou);
+        String location = getLDAPAttribute(certificatePrincipal, LOCATION);
+        LOGGER.debug("Location from certificate: ", location);
+
+        if(StringUtils.isBlank(ou)) {
+            throw new NoSuchAttributeException("No ou in dn, you may need to update your certificate: " + certificatePrincipal);
+        } else {
+            if(getAllAccessOu() != null && getAllAccessOu().equalsIgnoreCase(StringUtils.replace(ou, " ", ""))){
+                LOGGER.debug("Skipping environment validation, user ou matches {} ", getAllAccessOu());
+            } else {
+                //if dn not from allAccessOu, verify the location (l) field
+                //in the cert matches the configured environment
+                if(StringUtils.isBlank(location)) {
+                    throw new NoSuchAttributeException("No location in dn, you may need to update your certificate: " + certificatePrincipal);
+                } else if(!locationMatchesEnvironment(location)){
+                    throw new NoSuchAttributeException("Invalid location from dn, expected " + getEnvironment() + " but found l=" + location);
+                }
+            }
+        }
+    }
+
+    public String getLDAPAttribute(String certificatePrincipal, String searchAttribute) throws NamingException {
```

Comment:
Maybe move this method to LdapRolesProvider, since it doesn't really anything to do with environment.

Reply:
Environment doesn't know about any of the RolesProviders, I would like to keep it that way. Environment is already acting on DNs, so I don't think it is out of line to have this method here.